remote_shell_service.go

```diff
@@ -0,0 +1,222 @@
+		remoteSessions: map[uuid.UUID]ssh.Session{},
+	}
+}
+
+func (c *Console) WriteToStdIn(p []byte) (n int, err error) {
+	c.stdInLock.Lock()
+	n, err = c.stdInPipe.Write(p)
+	c.stdInLock.Unlock()
+
+	return n, err
+}
+
+func (c *Console) RegisterSession(id uuid.UUID, session ssh.Session) {
+	c.sessionLock.Lock()
+	c.remoteSessions[id] = session
+	c.sessionLock.Unlock()
+}
+
+func (c *Console) UnregisterSession(id uuid.UUID) {
+	c.sessionLock.Lock()
+	delete(c.remoteSessions, id)
+	c.sessionLock.Unlock()
+}
+
+func passwordHandler(ctx ssh.Context, password string, logger *zap.Logger) bool {
+	expectedPassword := os.Getenv("RCON_PASSWORD")
+	if expectedPassword == "" {
+		expectedPassword = "minecraft"
+	}
+
+	lengthComp := subtle.ConstantTimeEq(int32(len(password)), int32(len(expectedPassword)))
+	contentComp := subtle.ConstantTimeCompare([]byte(password), []byte(expectedPassword))
+	isValid := lengthComp == 1 && contentComp == 1
+	if !isValid {
+		logger.Warn(fmt.Sprintf("Remote console session rejected (%s/%s)", ctx.User(), ctx.RemoteAddr().String()))
+	}
+	return isValid
+}
+
+func handleSession(session ssh.Session, console *Console, logger *zap.Logger) {
+	// Setup state for the console session
+	sessionId := uuid.New()
+	_, _, isTty := session.Pty()
+	logger.Info(fmt.Sprintf("Remote console session accepted (%s/%s) isTTY: %t", session.User(), session.RemoteAddr().String(), isTty))
+	console.RegisterSession(sessionId, session)
+
+	lines := make(chan string)
+	clientExit := make(chan struct{})
+
+	go func() {
+		terminal := term.NewTerminal(session, "")
+		for {
+			line, err := terminal.ReadLine()
+			if err != nil {
+				if err != io.EOF {
+					logger.Error(fmt.Sprintf("Unable to read line from session (%s/%s)", session.User(), session.RemoteAddr().String()), zap.Error(err))
+				}
+				close(clientExit)
+				return
+			}
+
+			lines <- line
+		}
+	}()
+
+	shouldRun := true
+	for shouldRun {
+		select {
+		case line := <-lines:
+			lineBytes := []byte(fmt.Sprintf("%s\n", line))
+			_, err := console.WriteToStdIn(lineBytes)
+			if err != nil {
+				logger.Error(fmt.Sprintf("Session failed to write to stdin (%s/%s)", session.User(), session.RemoteAddr().String()), zap.Error(err))
+			}
+		case <-clientExit:
+			shouldRun = false
+		case <-session.Context().Done():
+			shouldRun = false
+		}
+	}
+
+	// Tear down the session
+	console.UnregisterSession(sessionId)
+	logger.Info(fmt.Sprintf("Remote console session disconnected (%s/%s)", session.User(), session.RemoteAddr().String()))
+}
+
+// Use os.Stdout for console.
+// There should only ever be one of these...
+func stdOutRoutine(stdOut io.Writer, console *Console, logger *zap.Logger) {
+	scanner := bufio.NewScanner(console.stdOutPipe)
+	for scanner.Scan() {
+		output := []byte(fmt.Sprintf("%s\n", scanner.Text()))
+		_, err := stdOut.Write(output)
+		if err != nil {
+			logger.Error("Failed to write to stdout")
+		}
+
+		for _, session := range console.remoteSessions {
```

Comment:
I'm not 100% sure this is safe without using the mutex in Go. I assume it isn't?

Reply:
Correct, you'll need to use acquire the session lock around this loop